File: backend/src/services/iih/IIHService.ts
import axios, { AxiosInstance } from 'axios';
import { logger } from '../../utils/logger';
import { config } from '../../config';
import { db } from '../../db/database';
import { broadcast } from '../../utils/websocket';
import { eventQueueService } from '../queue/EventQueueService';

export interface Equipment {
  id: string;
  name: string;
  description?: string;
  opcua_node_id: string;
  iih_asset_id: string;
  iih_variable_id: string;
  enabled: number;
  created_at: number;
  updated_at: number;
}

export interface CounterReading {
  value: number;
  unit: string;
  periodStart: string;
  periodEnd: string;
}

interface SettingRow {
  value: string;
}

function getEffectiveConfig(): { baseUrl: string; username: string; password: string; counterEndpoint: string } {
  const getSetting = (key: string): string | undefined => {
    const row = db.prepare('SELECT value FROM settings WHERE key = ?').get(key) as SettingRow | undefined;
    return row?.value;
  };

  return {
    baseUrl: getSetting('iihBaseUrl') || config.iih.baseUrl,
    username: getSetting('iihUsername') || config.iih.username,
    password: getSetting('iihPassword') || config.iih.password,
    counterEndpoint: getSetting('iihCounterEndpoint') || config.iih.counterEndpoint,
  };
}

function createAxiosInstance(cfg: ReturnType<typeof getEffectiveConfig>): AxiosInstance {
  return axios.create({
    baseURL: cfg.baseUrl,
    timeout: 30000,
    headers: { 'Content-Type': 'application/json' },
    auth: cfg.username ? { username: cfg.username, password: cfg.password } : undefined,
  });
}

interface IIHAggregatedValue {
  value: number;
  unit?: string;
  from?: string;
  to?: string;
}

interface IIHResponse {
  values?: IIHAggregatedValue[];
  data?: IIHAggregatedValue[];
}

export class IIHService {
  private lastStatus: 'connected' | 'disconnected' | 'error' | 'unknown' = 'unknown';

  async fetchCounterData(equipment: Equipment): Promise<CounterReading | null> {
    const cfg = getEffectiveConfig();
    const client = createAxiosInstance(cfg);

    const now = new Date();
    const periodEnd = now.toISOString();
    const periodStart = new Date(now.getTime() - 60 * 1000).toISOString();

    try {
      const response = await client.get<IIHResponse>(cfg.counterEndpoint, {
        params: {
          assetId: equipment.iih_asset_id,
          variableId: equipment.iih_variable_id,
          from: periodStart,
          to: periodEnd,
          aggregate: 'last',
        },
      });

      const data = response.data;
      const values = data.values || data.data || [];

      if (!values || values.length === 0) {
        logger.debug(`IIH: no counter data for equipment=${equipment.name}`);
        return null;
      }

      const latest = values[values.length - 1];
      return {
        value: typeof latest.value === 'number' ? latest.value : parseFloat(String(latest.value)),
        unit: latest.unit || 'units',
        periodStart: latest.from || periodStart,
        periodEnd: latest.to || periodEnd,
      };
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      logger.error(`IIH fetchCounterData failed for ${equipment.name}: ${message}`);
      return null;
    }
  }

  async pollAllEquipments(): Promise<void> {
    const equipments = db.prepare(`
      SELECT * FROM equipments WHERE enabled = 1
    `).all() as Equipment[];

    if (equipments.length === 0) {
      logger.debug('IIH poll: no enabled equipments');
      return;
    }

    logger.info(`IIH poll: fetching counter data for ${equipments.length} equipment(s)`);

    let successCount = 0;
    let errorCount = 0;

    for (const equipment of equipments) {
      try {
        const reading = await this.fetchCounterData(equipment);

        if (reading !== null) {
          const now = new Date().toISOString();
          eventQueueService.enqueue({
            type: 'COUNTER',
            equipment_id: equipment.id,
            equipment_name: equipment.name,
            payload: {
              equipmentId: equipment.id,
              equipmentName: equipment.name,
              iihAssetId: equipment.iih_asset_id,
              iihVariableId: equipment.iih_variable_id,
              counterValue: reading.value,
              unit: reading.unit,
              periodStart: reading.periodStart,
              periodEnd: reading.periodEnd,
              collectedAt: now,
            },
          });
          successCount++;
        }
      } catch (err) {
        errorCount++;
        logger.error(`IIH poll error for ${equipment.name}: ${err instanceof Error ? err.message : String(err)}`);
      }
    }

    const newStatus = errorCount === 0 ? 'connected' : (successCount === 0 ? 'error' : 'connected');

    if (newStatus !== this.lastStatus) {
      this.lastStatus = newStatus;
      broadcast('iih_status', { status: newStatus, successCount, errorCount });
    }

    logger.info(`IIH poll complete: success=${successCount} errors=${errorCount}`);
  }

  async testConnection(baseUrl: string, username?: string, password?: string, counterEndpoint?: string): Promise<{ success: boolean; message: string }> {
    try {
      const client = axios.create({
        baseURL: baseUrl,
        timeout: 10000,
        headers: { 'Content-Type': 'application/json' },
        auth: username ? { username, password: password || '' } : undefined,
      });

      const endpoint = counterEndpoint || config.iih.counterEndpoint;
      await client.get(endpoint, {
        params: { assetId: 'test', variableId: 'test', from: new Date().toISOString(), to: new Date().toISOString() },
      });

      return { success: true, message: 'IIH connection successful' };
    } catch (err) {
      if (axios.isAxiosError(err) && err.response) {
        // A response means the server is reachable (even if 4xx)
        if (err.response.status === 401) {
          return { success: false, message: 'IIH authentication failed: invalid credentials' };
        }
        if (err.response.status === 404) {
          return { success: true, message: `IIH server reachable but endpoint not found (${err.response.status})` };
        }
        return { success: true, message: `IIH server reachable (HTTP ${err.response.status})` };
      }
      const message = err instanceof Error ? err.message : String(err);
      return { success: false, message: `IIH connection failed: ${message}` };
    }
  }

  getStatus(): 'connected' | 'disconnected' | 'error' | 'unknown' {
    return this.lastStatus;
  }
}

export const iihService = new IIHService();

File: backend/services/iih/iih_service.py
from datetime import datetime, timedelta, timezone
from typing import Any

import httpx

from config import config
from db import get_db
from services.queue.event_queue_service import event_queue_service
from utils.logger import logger
from utils.websocket import broadcast


def _get_effective_config() -> dict:
    db = get_db()

    def _get(key: str) -> str:
        row = db.execute("SELECT value FROM settings WHERE key = ?", (key,)).fetchone()
        return row["value"] if row else ""

    return {
        "base_url": _get("iihBaseUrl") or config.iih.base_url,
        "username": _get("iihUsername") or config.iih.username,
        "password": _get("iihPassword") or config.iih.password,
        "counter_endpoint": _get("iihCounterEndpoint") or config.iih.counter_endpoint,
    }


class IIHService:
    def __init__(self) -> None:
        self._last_status: str = "unknown"

    async def fetch_counter_data(self, equipment: dict) -> dict | None:
        cfg = _get_effective_config()
        now = datetime.now(timezone.utc)
        period_end = now.isoformat()
        period_start = (now - timedelta(seconds=60)).isoformat()

        auth = (cfg["username"], cfg["password"]) if cfg["username"] else None
        try:
            async with httpx.AsyncClient(
                base_url=cfg["base_url"],
                auth=auth,
                timeout=30.0,
                headers={"Content-Type": "application/json"},
            ) as client:
                response = await client.get(
                    cfg["counter_endpoint"],
                    params={
                        "assetId": equipment["iih_asset_id"],
                        "variableId": equipment["iih_variable_id"],
                        "from": period_start,
                        "to": period_end,
                        "aggregate": "last",
                    },
                )
                response.raise_for_status()
                data = response.json()

            values = data.get("values") or data.get("data") or []
            if not values:
                logger.debug(f"IIH: no counter data for equipment={equipment['name']}")
                return None

            latest = values[-1]
            raw_val = latest.get("value", 0)
            return {
                "value": float(raw_val) if not isinstance(raw_val, float) else raw_val,
                "unit": latest.get("unit", "units"),
                "period_start": latest.get("from", period_start),
                "period_end": latest.get("to", period_end),
            }
        except Exception as exc:
            logger.error(f"IIH fetchCounterData failed for {equipment['name']}: {exc}")
            return None

    async def poll_all_equipments(self) -> None:
        db = get_db()
        rows = db.execute("SELECT * FROM equipments WHERE enabled = 1").fetchall()
        equipments = [dict(r) for r in rows]

        if not equipments:
            logger.debug("IIH poll: no enabled equipments")
            return

        logger.info(f"IIH poll: fetching counter data for {len(equipments)} equipment(s)")
        success_count = 0
        error_count = 0

        for equipment in equipments:
            try:
                reading = await self.fetch_counter_data(equipment)
                if reading is not None:
                    now = datetime.now(timezone.utc).isoformat()
                    event_queue_service.enqueue(
                        event_type="COUNTER",
                        equipment_id=equipment["id"],
                        equipment_name=equipment["name"],
                        payload={
                            "equipmentId": equipment["id"],
                            "equipmentName": equipment["name"],
                            "iihAssetId": equipment["iih_asset_id"],
                            "iihVariableId": equipment["iih_variable_id"],
                            "counterValue": reading["value"],
                            "unit": reading["unit"],
                            "periodStart": reading["period_start"],
                            "periodEnd": reading["period_end"],
                            "collectedAt": now,
                        },
                    )
                    success_count += 1
            except Exception as exc:
                error_count += 1
                logger.error(f"IIH poll error for {equipment['name']}: {exc}")

        new_status = "connected" if error_count == 0 else ("error" if success_count == 0 else "connected")
        if new_status != self._last_status:
            self._last_status = new_status
            await broadcast("iih_status", {"status": new_status, "successCount": success_count, "errorCount": error_count})

        logger.info(f"IIH poll complete: success={success_count} errors={error_count}")

    async def test_connection(
        self, base_url: str, username: str = "", password: str = "",
        counter_endpoint: str = ""
    ) -> dict[str, Any]:
        endpoint = counter_endpoint or config.iih.counter_endpoint
        auth = (username, password) if username else None
        try:
            async with httpx.AsyncClient(base_url=base_url, auth=auth, timeout=10.0) as client:
                response = await client.get(
                    endpoint,
                    params={"assetId": "test", "variableId": "test",
                            "from": datetime.now(timezone.utc).isoformat(),
                            "to": datetime.now(timezone.utc).isoformat()},
                )
            if response.status_code == 401:
                return {"success": False, "message": "IIH authentication failed: invalid credentials"}
            if response.status_code == 404:
                return {"success": True, "message": f"IIH server reachable but endpoint not found (404)"}
            return {"success": True, "message": f"IIH server reachable (HTTP {response.status_code})"}
        except Exception as exc:
            return {"success": False, "message": f"IIH connection failed: {exc}"}

    def get_status(self) -> str:
        return self._last_status


iih_service = IIHService()
